### ISSUE TITLE
Support login via JWT(?)

### ISSUE THREAD

**Issue by Johennes** [Dominaezzz/matrix-kt#3]
Synapse supports a non-standard login type based on JSON Web Tokens: https://matrix-org.github.io/synapse/latest/jwt.html

Would it be possible to include this in matrix-kt? Happy to help if that would be appreciated. 🙂

**Comment by Dominaezzz** [comment]
`matrix-kt` is designed to allow non-standard stuff like this be plugged in without directly including it. (Although I do plan to have a "synapse extensions" library at some point)

You just need to make a new login type by extending [`Login.Body`](https://github.com/Dominaezzz/matrix-kt/blob/3b0af7807b400775d73fde48d401d3d7098d4d73/client/src/main/kotlin/io/github/matrixkt/api/Login.kt#L32) like so.

```kotlin
@SerialName("org.matrix.login.jwt")
@Serializable
data class JsonWebToken(
    val token: String,

    override val identifier: UserIdentifier? = null,

    @SerialName("device_id")
    override val deviceId: String? = null,

    @SerialName("initial_device_display_name")
    override val initialDeviceDisplayName: String? = null
) : Login.Body()
```

Then you register this new type in a new serializer module.
```kotlin
val MyJson = Json(MatrixJson) {
    serializersModule += SerializersModule {
        polymorphic(Login.Body::class, JsonWebToken::class, JsonWebToken.serializer())

        // Workaround for ktor bug, will be fixed in Ktor 2.0.0
        contextual(JsonWebToken::class, Login.Body.serializer() as KSerializer<JsonWebToken>)
    }
}
```

You can then use your custom `Json` instance when creating your ktor client.
```kotlin
val client = HttpClient { MatrixConfig(json = MyJson) }

client.rpc(Login(Login.Url(), JsonWebToken("<secret token>")))
```

Let me know if this works for you or if you have issues.

**Comment by Johennes** [comment]
Oh nice! I didn't realize I could extend it like that. That seems like it should work for my use case. Thanks a lot!
